admin/sections/knowledge.tsx

```diff
@@ -16,79 +11,129 @@ import {
+      return prevFiles.filter(
+        (f) => !toRemoveFiles.find((rf) => rf.name === f.name),
+      );
+    });
+    if (setSubmit) {
+      setIsSubmitting(false);
     }
   }
 
-  async function handleAddFiles(addingFiles: any[]) {
-    for (const file of addingFiles) {
-      // Add the file to list files with uploading status
-      const fileObj = {
-        name: file.name,
-        status: "uploading" as FileStatus,
-      };
-      setFiles((prevFiles) => [...prevFiles, fileObj]);
-      // Upload the file to the server
+  async function handleAddFiles(addingFiles: globalThis.File[]) {
+    const toUploadFiles: File[] = addingFiles.map((file) => ({
+      name: file.name,
+      status: "toUpload",
+      blob: file,
+    }));
+    setFiles((prevFiles) => [...prevFiles, ...toUploadFiles]);
+    return toUploadFiles;
+  }
+
+  async function handleUploadFile(toUploadFiles?: File[]) {
+    if (!toUploadFiles) {
+      toUploadFiles = files.filter((file) => file.status === "toUpload");
+    }
+    console.log("Uploading files:", toUploadFiles);
+
+    // Change to for loop to upload files one by one
+    for (const file of toUploadFiles) {
       const formData = new FormData();
-      formData.append("file", file);
+      formData.append("file", file.blob as Blob);
       try {
         await uploadFile(formData);
         setFiles((prevFiles) => {
-          return prevFiles.map(updateStatus(fileObj.name, "uploaded"));
+          return prevFiles.map((f) => {
+            if (f.name === file.name) {
+              return { ...f, status: "uploaded" };
+            }
+            return f;
+          });
         });
       } catch (err: unknown) {
-        // Remove the file from the list
-        setFiles((prevFiles) => {
-          const filteredFiles = prevFiles.filter(
-            (f) => f.name !== fileObj.name,
-          );
-          return filteredFiles;
-        });
-        // Show a error toast
         console.error(
           "Failed to upload the file:",
           file.name,
           (err as Error)?.message,
         );
         toast({
-          className: cn(
-            "top-0 right-0 flex fixed md:max-w-[420px] md:top-4 md:right-4 text-red-500",
-          ),
           title: "Failed to upload the file: " + file.name + "!",
+          variant: "destructive",
         });
+        setFiles((prevFiles) => {
+          return prevFiles.map((f) => {
+            if (f.name === file.name) {
+              return { ...f, status: "failed" };
+            }
+            return f;
+          });
+        });
+        await handleRemoveFiles([file], false);
       }
     }
   }
 
+  async function processUpload(blobFiles: globalThis.File[]) {
+    // Ignore duplicate files
+    const duplicateFiles = files.filter((file) =>
+      blobFiles.find((f) => f.name === file.name),
+    );
+    if (duplicateFiles.length > 0) {
+      toast({
+        title: `Duplicate files (${duplicateFiles.map((f) => f.name).join(", ")})!`,
```

Comment:
we just continue with the upload and ignore duplicate files, right? then how about informing the user? e.g.
```suggestion
        title: `Ignoring duplicate files (${duplicateFiles.map((f) => f.name).join(", ")})!`,
```